### PR TITLE
Lazy source

### PR DESCRIPTION
**Pull request by johanandren**
Spawned from @drewhk s comment about Source.lazy in #21025 

Useful together with `orElse(fallBack)` to not materialize fallback unless needed.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/LazySource.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/LazySource.scala]: please be explicit about `() => Source`

File: akka-stream/src/main/scala/akka/stream/impl/LazySource.scala
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/LazySource.scala]: I would alternatively recommend using SubSink. You create a `SubSink` via `SubSinkInlet`, attach it to the source created by the factory method. Then you can use the usual grab/pull style coding there. 

The benefit compared to the current approach is that it would not break fusing and would be able to use the internal queue instead of the actor mailbox instead which is one magnitude faster AFAIK.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazySourceSpec.scala
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazySourceSpec.scala]: isn't this assuming too much of when the future is completed?
use `Await`?

Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazySourceSpec.scala]: True, may be a race. Will fix.

**Comment by ktoso** [comment]
LGTM, I'm not sure about the TODO you left there, can stay as is.
If additional LGTM comes we could merge this ;)

**Comment by ktoso** [comment]
PLS BUILD

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
Meh, first unrelated fail and now a7 not being friends.

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
I guess it should inherit the attributes just like how it was done in #21743

**Comment by ktoso** [comment]
> I guess it should inherit the attributes just like how it was done in #21743

Yeah I think so - seems that's the last fixup to make here though. I'd love to see this merged, already telling people we have it because it was around as PR for so long I thought it's in 😉

File: akka-stream/src/main/scala/akka/stream/impl/LazySource.scala
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/LazySource.scala]: yes, it should. Also, a postStop() hook which as a last resort tries to fail the Promise if it is not yet closed.

**Review by drewhk — APPROVED** [review]
LGTM, after small fixes

**Comment by johanandren** [comment]
Before it can be merged: needs doc mention I guess, does it need factory methods as well?

**Comment by johanandren** [comment]
Refs #21863

**Comment by johanandren** [comment]
Unrelated test failures in compressionspec.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: in javadsl it could be `lazy` btw hm

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: The javadsl Source class is written in Scala, it could if it was Java.

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Here's some scala magic to make it possible:

```
def `lazy` = ???
```

😎

**Review by ktoso — APPROVED** [review]
LGTM, I really want this one ;)

**Comment by johanandren** [comment]
Still the compression spec failing ;(

**Comment by johanandren** [comment]
pls build

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**
